README
Sat Feb 28 09:33:05 2009 Robert Heller <heller@deepsoft.com>

This is a port of Facebook's API that I translated from php to Tcl,
using SNIT as the OO framework and using Don Libes's cgi.tcl package to
implement the low-level CGI/HTML layer.  The API has been lightly
tested by using it for a simple Facebook application that I wrote.

The source tree is organized like this:

BuildScripts -- This directory contains various scripts / StarKits used
		to build the package.

Common       -- This directory contains the distributed library scripts.
		There are two: Facebook.tcl, which is a port of
		facebook.php and facebookapi_php5_restlib.php and
		cgi.tcl, which is Don Libes's Tcl code for writing CGI
		scripts in Tcl.

Doc	     -- Contains the documentation for the package.

Doc/Internals -- Contains the LaTeX source for the internals
		documentation.  Actually, most of the actual
		documentation is in the source code.  This directory
		contains the build logical to extract the documentation
		from the source code and product PDF and HTML manuals.

Doc/Support  -- Contains support for the internals documentation
		extraction.

HelloUser    -- Contains an example application script.

To 'build' the package you will need:

Tcl installed (8.4 or later).
Tcllib installed (includes the snit, md5, and uri packages).
TclXML installed
At least a native tclkit (download one from http://www.equi4.com/tclkit.html)
SDX, the Starkit/Starpack utility (download from http://www.equi4.com/sdx.html)
LaTeX and tex4ht are needed to build the documentation.

If your web server has Tcl, Tcllib, and TclXML installed, you don't
actually have to build anything.  Just make a directory named
TclIncludes that is parallel to your websites DOCUMENT_ROOT and put all
of the Tcl files in Common in this directory (there are three files:
Facebook.tcl, cgi.tcl, and pkgIndex.tcl), and then put your application
script (adapted from the HelloUser.tcl file in HelloUser) someplace
where a CGI script would go -- either a cgi-bin directory (a directory
that is a ScriptAlias w/ Options ExecCGI OR a directory with a
'AddHandler cgi-script .cgi' directive.  Rename it with a proper
extension (eg .cgi) and make sure it is executable (chmod'ed with +x)
and you should be all set.  Make sure that your script file starts with
a line starting with '#!' and containg the proper path to your installed
tclsh executable.  You can use pgtslsh if you want to use PostgreSQL.


ChangeLog
Tue Feb 16 16:33:25 2010

   Release 1.6.3: Various minor fixes.
		Added additional (new) methods.

Thu Feb  4 16:30:43 2010

   Release 1.4.2: Various minor fixes.
		Second 1.5 release candidate *untested*.

Thu Feb  4 13:43:54 2010

   Release 1.4.1: Changed to support both Don Libes's cgi.tcl AND Rivet.
		This is a 1.5 release candidate *untested*.
		

Thu Feb  4 12:30:52 2010

    Release 1.4: Small fix: 
		$generate_session_secret => $options(-generate_session_secret)
		in method validate_fb_params.

Sat Feb 28 10:11:09 2009

    Release 1.1: Minor Documentation updates (README, AUTHORS, ChangeLog)


Fri Feb 27 04:12:44 PM EST 2009

    Initial 1.0 release.


HelloUser/HelloUser.tcl
#!/usr/bin/tclsh
#* 
#* ------------------------------------------------------------------
#* HelloUser.tcl - Hello User example Facebook application
#* Created by Robert Heller on Wed Feb 25 13:30:33 2009
#* ------------------------------------------------------------------
#* ------------------------------------------------------------------
#* Contents:
#* ------------------------------------------------------------------
#*  
#*     Facebook API in Tcl Project
#*     Copyright (C) 2009  Robert Heller D/B/A Deepwoods Software
#* 			51 Locke Hill Road
#* 			Wendell, MA 01379-9728
#* 
#*     This program is free software; you can redistribute it and/or modify
#*     it under the terms of the GNU General Public License as published by
#*     the Free Software Foundation; either version 2 of the License, or
#*     (at your option) any later version.
#* 
#*     This program is distributed in the hope that it will be useful,
#*     but WITHOUT ANY WARRANTY; without even the implied warranty of
#*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#*     GNU General Public License for more details.
#* 
#*     You should have received a copy of the GNU General Public License
#*     along with this program; if not, write to the Free Software
#*     Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#* 
#*  
#* 

#@Chapter:HelloUser.tcl -- Sample Application
#@Label:HelloUser.tcl
#$Id: HelloUser.tcl 799 2010-02-04 19:11:13Z heller $
# This is a trivial application and is a minor variation of a classic 
# 'Hello World' program.

# Check to see how we were started up -- this would either be as a bare
# Tcl script or as a Starkit or Starpack.  If as a Tcl script, we need to
# find our support code (Facebook.tcl and cgi.tcl, plus support packages).
# As a Tcl script, we are presumed to be in a cgi-bin directory (or somewhere
# with either 'Options ExecCGI' and/or a 'AddHandler cgi-script .cgi' directive
# We will assume that there is a TclIncludes directory adjacent to the 
# DOCUMENT_ROOT -- that is if the the DOCUMENT_ROOT is '...PATH/WWW', then the 
# Tcl library code is in  '...PATH/TclIncludes'.  We further assume that either 
# the other support packages (md5, snit, uri, and tclxml) are installed under 
# there OR they are installed somewhere else under the standard places 
# (/usr/lib, /usr/share/tcl8.<mumble>, etc.).
#
# If we were started as a Starkit or Starpack, all of the support code is
# embedded in the Starkit or Starpack and there is no need to mess with
# auto_path.

if {![namespace exists ::starkit]} {
  lappend auto_path [file join [file dirname $::env(DOCUMENT_ROOT)] \
			       TclIncludes]
}

# Now load Facebook & cgi code, along with the other support packages.

package require Facebook

namespace eval HelloUser {
  # Namespace containing the HelloUser application code.

  variable AdminEMailAddress myname@my.domain.name;# Admin E-Mail address
  variable APIKEY 123456789abcdef0123456789abcdef0;# Your API KEY
  variable SECRET 0fedcba9876543210fedcba987654321;# Your API Secret
  variable CGI_DebugFlags -off;# Change this to -on for debugging

  variable DOCUMENT_ROOT;#	The document root.
  variable SERVER_NAME;#	The hostname of the server.
  variable SCRIPT_NAME;#	The name of the CGI script.
  variable REQUEST_URI;#	The request URI.
  variable MyName;#		The base name this script was called under.
  #				We can share code for different script
  #				functions by making multiple hard links
  #				under different names.

  # Set up cgi environment settings
  if {[catch {
    set DOCUMENT_ROOT $::env(DOCUMENT_ROOT)
    set SERVER_NAME   $::env(SERVER_NAME)
    set SCRIPT_NAME   $::env(SCRIPT_NAME)
    set REQUEST_URI   $::env(REQUEST_URI)
    } error]} {
    error "$argv0: Not started as a CGI Script!  Opps! Error was: $error"
  }
  set MyName [file rootname [file tail $SCRIPT_NAME]]
  set cb [file dirname $REQUEST_URI]
  if {"$cb" eq "/" || "$cb" eq "."} {set cb {}}
  ::cgi_root http://$SERVER_NAME$cb
  ::cgi_name $MyName
  ::cgi_admin_mail_addr $AdminEMailAddress
  ::cgi_debug $CGI_DebugFlags

  # Start of application code proper.
  # Wrap it in a cgi_eval -- this catches errors and ensures that proper http
  # headers (and html) is sent in the event of an error
  ::cgi_eval {
    ::cgi_input;#	Process CGI inputs

    # Get a Facebook object (grabs Facebook parameters
    variable fb [::Facebook::Facebook %%AUTO%% -api_key $APIKEY -secret $SECRET]

    # Require a frame and require login
    $fb require_frame
    variable UserId [$fb require_login]

    # Send HTTP headers
    ::cgi_http_head
    # Generate some HTML / FBML code.
    ::cgi_division {style="padding: 10px;"} {
      h2 "Hi <fb:name firstnameonly=\"true\" uid=\"$UserId\" useyou=\"false\"/>!"
      br
      if {![$fb added-p]} {
        cgi_put "[url {Put the Hello User application in your profile!} [$fb get_add_url]]"
      } else {
        if {![$fb profile_setFBML resultArray {} $UserId \
	  "Hi <fb:name firstnameonly=\"true\" uid=\"$UserId\" useyou=\"false\"/>!" \
	  {} {} "Hi <fb:name firstnameonly=\"true\" uid=\"$UserId\" useyou=\"false\"/>!"]} {
	  error "profile_setFBML: [$fb getErrorCode] [$fb getErrorMessage]"
	}
      }
    }
  }
}

  
